gapic-common/lib/gapic/grpc/errors.rb

```diff
@@ -0,0 +1,63 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "googleauth"
+require "gapic/common/error"
+
+module Gapic
+  # It's called GRPCName because otherwise I get naming errors
+  # where everything referencing GRPC::Anything will try and fail to 
+  # find it here. 
+  # I've elected to use this as a stopgap and do namespace disambiguation later
+  module GRPCName
+    # Just a baseline class
+    class GRPCError < Gapic::Common::Error # or maybe GRPC::BadStatus?
+    end
+
+    ##
+    # An error class to represent the Authorization Error.
+    # Needed since GRPC layer wraps auth plugin errors in ::GRPC::Unavailable.
+    #
+    class AuthorizationError < Gapic::GRPCName::GRPCError
+      ##
+      # Returns a code that can be tested against a grpc retry policy
+      #
+      def code
+        ::GRPC::Core::StatusCodes::UNAUTHENTICATED
+      end
+    end
+
+    ##
+    # An error class that wraps both GRPC's DeadlineExceeded and a root cause
+    # error that was being retried until the deadline was exceeded (if any).
+    # Needed since GRPC's DeadlineExceededError does not contain a reference
+    # to a root cause error.
```

Comment:
So am I understanding correctly that the intent is that `cause` will be the underlying GRPC DeadlineExceededError, and `root_cause` will be the original error that caused the retries? And that's why we have two separate "causes"? (If so, I think that's fine, but we'll need to document it carefully.)

Reply:
yes, you got it right.